### ISSUE TITLE
Try to convert string value in Double/Float decoders

### ISSUE THREAD

**Issue by knutwalker** [travisbrown/circe#173]
From the [Gitter conversation](https://gitter.im/travisbrown/circe?at=56a69f40c54bc2bf180bfcd6).

The Decoders for `Int`, `Long` and other integral types try to convert a string into their number type.

```scala
import io.circe._, parse._, cats.data.Xor
scala> val Xor.Right(i) = decode[Int]("\"42\"")
i: Int = 42
```

The decoder for `Double` and `Float` don't try this conversion because of this:

```
scala> "Infinity".toDouble
res1: Double = Infinity
```

However, we could use `JsonNumber.fromString` and `_.toDouble` to try to safely convert a String into a Double, for example:

```scala
scala> val doubleWithFromString = Decoder.instance(c => c.as[String].flatMap(s => Xor.fromOption(JsonNumber.fromString(s).map(_.toDouble), DecodingFailure("Double", c.history))))
doubleWithFromString: io.circe.Decoder[Double] = io.circe.Decoder$$anon$8@47016c8b

scala> val Xor.Right(d) = decode("\"13.37\"")(doubleWithFromString)
d: Double = 13.37

scala> val Xor.Left(e) = decode("\"Infinity\"")(doubleWithFromString)
e: io.circe.Error = io.circe.DecodingFailure: Double

scala> val Xor.Left(e) = decode("\"NaN\"")(doubleWithFromString)
e: io.circe.Error = io.circe.DecodingFailure: Double
```

If `"Infinty"` and `"NaN"` are valid to be parsed as Double, one can fallback to using a `Decoder[String].map(_.toDouble)` approach.

**Comment by MartinSeeler** [comment]
I think this is a good idea!

**Comment by travisbrown** [comment]
Fixed in #189.
